File: sort/radixSort/sort_kotlin/src/Lblbc.kt
/**
 * 厦门大学计算机专业 | 前华为工程师
 * 专注《零基础学编程系列》  http://lblbc.cn/blog
 * 包含：Java | 安卓 | 前端 | Flutter | iOS | 小程序 | 鸿蒙
 * 公众号：蓝不蓝编程
 */

fun main() {
    val array = intArrayOf(2, 1, 5, 4, 3)
    println("排序前")
    printArray(array)
    sort(array)
    println("排序后")
    printArray(array)
}

private fun sort(array: IntArray) {
    val max = getMaxValue(array)
    var exp = 1
    while (max / exp > 0) {
        countSort(array, exp)
        exp *= 10
    }
}

private fun countSort(array: IntArray, exp: Int) {
    val tmpArr = IntArray(array.size)
    val bucketArr = IntArray(10)
    for (i in array.indices) bucketArr[array[i] / exp % 10]++
    for (i in 1..9) bucketArr[i] += bucketArr[i - 1]
    for (i in array.indices.reversed()) {
        tmpArr[bucketArr[array[i] / exp % 10] - 1] = array[i]
        bucketArr[array[i] / exp % 10]--
    }
    for (i in array.indices) {
        array[i] = tmpArr[i]
    }
}

private fun getMaxValue(array: IntArray): Int {
    var max = array[0]
    for (i in 1 until array.size) {
        if (array[i] > max) {
            max = array[i]
        }
    }
    return max
}

private fun printArray(array: IntArray) {
    for (i in array.indices) {
        print(array[i].toString() + " ")
    }
    println()
}

File: sort/radixSort/sort_C++/Lblbc.cpp
/// 厦门大学计算机专业 | 前华为工程师
/// 专注《零基础学编程系列》  http://lblbc.cn/blog
/// 包含：Java | 安卓 | 前端 | Flutter | iOS | 小程序 | 鸿蒙
/// 公众号：蓝不蓝编程

#include <iostream>
using namespace std;

void printArray(int* array, int count)
{
	for (int i = 0; i < count; i++)
	{
		cout << array[i] << " ";
	}
}

void sort(int *arr, int n)
{
	int max = arr[0];
	int base = 1;

	for (int i = 0; i < n; i++)
	{
		if (arr[i] > max)
		{
			max = arr[i];
		}
	}
	int *tmpArr = new int[n];
	while (max / base > 0)
	{
		int bucket[10] = {0};
		for (int i = 0; i < n; i++)
		{
			bucket[arr[i] / base % 10]++;
		}
		for (int i = 1; i < 10; i++)
		{
			bucket[i] += bucket[i - 1];
		}
		for (int i = n - 1; i >= 0; i--)
		{
			tmpArr[bucket[arr[i] / base % 10] - 1] = arr[i];
			bucket[arr[i] / base % 10]--;
		}
		for (int i = 0; i < n; i++)
		{
			arr[i] = tmpArr[i];
		}
		base *= 10;
	}
	free(tmpArr);
}

int main()
{
	int array[5] = { 2, 1, 5, 4, 3 };
	int count = sizeof(array) / sizeof(int);
	printf("排序前:");
	printArray(array, count);

	sort(array, count);

	printf("排序后:");
	printArray(array, count);
	return 0;
}

